data/model/pull_statistics.py

```diff
@@ -0,0 +1,236 @@
+"""
+Business logic for pull statistics operations.
+
+This module provides functions for managing pull statistics data:
+- Bulk upsert operations for tag and manifest statistics
+- Helper functions for individual record operations
+- Query functions for retrieving statistics
+"""
+import logging
+from datetime import datetime
+from typing import Dict, List, Optional
+
+from peewee import IntegrityError
+
+from data.database import (
+    ManifestPullStatistics,
+    Repository,
+    TagPullStatistics,
+    db_transaction,
+)
+from data.model import DataModelException
+
+logger = logging.getLogger(__name__)
+
+
+class PullStatisticsException(DataModelException):
+    """
+    Exception raised when pull statistics operations fail.
+    """
+
+    pass
+
+
+def bulk_upsert_tag_statistics(tag_updates: List[Dict]) -> int:
+    """
+    Upsert tag pull statistics in bulk.
+
+    Args:
+        tag_updates: List of dictionaries with tag statistics data:
+            [
+                {
+                    'repository_id': 123,
+                    'tag_name': 'latest',
+                    'pull_count': 5,
+                    'last_pull_timestamp': '2024-01-01T12:00:00Z',
+                    'manifest_digest': 'sha256:abc...'
+                }
+            ]
+
+    Returns:
+        Number of rows affected
+
+    Raises:
+        PullStatisticsException: If bulk operation fails
+    """
+    if not tag_updates:
+        return 0
+
+    rows_affected = 0
+
+    with db_transaction():
+        for update in tag_updates:
+            try:
+                # Parse timestamp - handle both datetime objects and ISO strings
+                last_pull_timestamp = update["last_pull_timestamp"]
+                if isinstance(last_pull_timestamp, datetime):
+                    last_pull = last_pull_timestamp
+                elif isinstance(last_pull_timestamp, str):
+                    last_pull = datetime.fromisoformat(last_pull_timestamp.replace("Z", "+00:00"))
+                else:
+                    raise ValueError(f"Invalid timestamp format: {last_pull_timestamp}")
+
+                # Try to get existing record
+                try:
+                    existing = TagPullStatistics.get(
+                        TagPullStatistics.repository == update["repository_id"],
+                        TagPullStatistics.tag_name == update["tag_name"],
+                    )
+
+                    # Update existing record
+                    existing.tag_pull_count += update["pull_count"]
+                    existing.last_tag_pull_date = max(existing.last_tag_pull_date, last_pull)
+                    existing.current_manifest_digest = update["manifest_digest"]
+                    try:
+                        existing.save()
+                        rows_affected += 1
+                    except Exception as save_error:
```

Comment:
nit: keeping it consistent with rest of the code
```suggestion
                    except Exception as save_error:
                        logger.error(f"Failed to save existing tag pull statistics record: {save_error}")
                        raise
```